### DISCUSSION TITLE
Template no longer runs function since latest update?

### DISCUSSION THREAD

**Discussion by fluckx-cs** [JanDeDobbeleer/oh-my-posh#5558] [General]
Hello,

In a previous version ( unfortunately I don't know which one ) the `template` parsed the string so I could use a bash function inside of it. I used this to show my aws_vault profile and then show me a localized expiry date for the session so I don't start any longer running things on AWS ( like with terraform ) 2 minutes before my session expires.

The segment looks like this:

```
{
          "type": "aws",
          "style": "powerline",
          "foreground": "#fcaa67",
          "background": "#a53f2b",
          "powerline_symbol": "\ue0b0",
          "template": "{{ if .Profile }} \uF0C2  {{.Profile}} {{ if .Env.AWS_CREDENTIAL_EXPIRATION }} [  $(get_awsvault_expiry '{{ .Env.AWS_CREDENTIAL_EXPIRATION }}' ) ] {{ end }} {{ end }}"
}
```

In the past it would look like this:

```
sandbox  [  2024-01-01 12:00:00 ]
```

With the latest version ( I updated monday ) It looks like this:

```
sandbox  [  $(get_awsvault_expiry '2024-09-04T15:00:59Z' ) ]
```

Did something change that I can no longer run this? Is there a better way to achieve this?
The `get_aws_vault_expiry` just converts the timestamp from UTC to my local time.

**Comment by fluckx-cs** [comment]
Update:

Colleague of mine is on 23.6.3 and it still works on that version.
I am on 23.10.1 and it no longer works there ( behaviour described above ).

**Reply by JanDeDobbeleer** [reply]
You need to set `interactive` to true on that segment. The fact that it worked was actually a bug.

**Reply by fluckx-cs** [reply]
Hi Jan,

Thanks for the reply. I tried that earlier, but it showed the same results.
Just confirmed that it isn't parsing running the command with `interactive` set to `true`.

Am I overlooking something?


```
{
  "type": "aws",
  "style": "powerline",
  "foreground": "#fcaa67",
  "background": "#a53f2b",
  "powerline_symbol": "\ue0b0",
  "interactive": true,
  "template": "{{ if .Env.AWS_VAULT }} \uF0C2  {{.Env.AWS_VAULT}} {{ if .Env.AWS_CREDENTIAL_EXPIRATION }} :: $(get_awsvault_expiry '{{ .Env.AWS_CREDENTIAL_EXPIRATION }}' ) {{ end }} {{ end }}"
}
```

Result:

```
 sandbox  :: $(get_awsvault_expiry '2024-09-04T15:00:59Z' )
```

Edit: Template string wasn't fully copied for some reason

**Reply by JanDeDobbeleer** [reply]
I will try to debug this tomorrow.

**Comment by lewis-yeung** [comment]
@fluckx-cs

> Did something change that I can no longer run this?

Yes, but not really. We've disabled expansions like command substitution for a prompt string in Bash/Zsh. Two reasons:

- They can produce unexpected results.
- With the exception of Bash/Zsh that interprets sequences such as `$var_foo` and `$(func_bar baz)` in prompt controlling by default, other shells treat any strings literally, thus breaking cross-shell consistency which is a goal of Oh My Posh.

> Is there a better way to achieve this? The `get_aws_vault_expiry` just converts the timestamp from UTC to my local time.

Of course. If the format of your `AWS_CREDENTIAL_EXPIRATION` environment variable never changes, you can just use a template like this:

```json
"template": "{{ if .Env.AWS_VAULT }} \uF0C2  {{.Env.AWS_VAULT}} {{ if .Env.AWS_CREDENTIAL_EXPIRATION }} :: {{ .Env.AWS_CREDENTIAL_EXPIRATION | toDate `2006-01-02T15:04:05Z0700` | date `2006-01-02 15:04:05` }} {{ end }} {{ end }}"
```

**Reply by JanDeDobbeleer** [reply]
@lewis-yeung that does imply I need to remove `interactive`, I'll do that in #3040

**Reply by lewis-yeung** [reply]
@JanDeDobbeleer The `interactive` works in another way, at least it does not break anything except #5455. Removing it can break a prompt since some Bash/Zsh users get used to it.

**Reply by JanDeDobbeleer** [reply]
You're right I'm confused. That's for prompt expansion, not interpretation.

**Comment by fluckx-cs** [comment]
Hi @JanDeDobbeleer ,

The ENV var is always in the same format so that will work for printing the date.
I'll see if I can get it to convert the timestamp from UTC to CET.

Thanks for the assistance!

**Reply by fluckx-cs** [reply]
For clarity:

the env var always contains a timestamp like this: `2024-09-04T15:00:59Z`.

**Reply by lewis-yeung** [reply]
> I'll see if I can get it to convert the timestamp from UTC to CET.


@fluckx-cs That's easy with the help of the template function `dateInZone`:

```
{{ dateInZone `2006-01-02 15:04:05` (.Env.AWS_CREDENTIAL_EXPIRATION | toDate `2006-01-02T15:04:05Z0700`) `CET` }}
```

**Reply by fluckx-cs** [reply]
Hi Lewis,

Thanks for the help! Works like a charm.
I'm still relatively new with the oh-my-posh and had found the dateInZone, but hadn't managed to do this yet.

Thanks :)

Another lesson for the road!

**Comment by fluckx-cs** [comment]
Thank you both for helping!
